onnxruntime/contrib_ops/webgpu/bert/multihead_attention.cc

```diff
@@ -0,0 +1,506 @@
+  shader.AddInput("qkv_input", ShaderVariable::UseUniform);
+  const auto& qkv_output = shader.AddOutput("qkv_output", ShaderVariable::UseUniform |
+      ShaderVariable::UseOffsetToIndices);
+
+  if (has_bias_) {
+    shader.AddInput("bias", ShaderVariable::UseUniform);
+  }
+
+  shader.SetMainFunctionBody(shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.data_size"),
+      "let output_indices = ", qkv_output.OffsetToIndices("global_idx"), ";\n",
+      "let input_offset_idx = output_indices[0] * uniforms.batch_offset + output_indices[1] * ",
+      "uniforms.head_offset + output_indices[2] * uniforms.sequence_offset + output_indices[3];\n",
+      has_bias_ ? "let bias_offset_idx = (input_offset_idx % uniforms.sequence_offset) + uniforms.bias_offset;\n" : "",
+      "qkv_output[global_idx] = qkv_input[input_offset_idx]",
+      has_bias_ ? " + bias[bias_offset_idx];\n" : ";\n");
+
+  return Status::OK();
+}
+
+Status TransferBSDToBNSH(onnxruntime::webgpu::ComputeContext& context, int num_heads, int sequence_length,
+    int head_size, const Tensor* input_tensor, const Tensor* bias, int bias_offset, Tensor* output_tensor) {
+  assert(input_tensor->Shape().GetDims().size() == 3);
+  assert(output_tensor->Shape().GetDims().size() == 4);
+
+  uint32_t data_size = SafeInt<uint32_t>(output_tensor->Shape().Size());
+  const int batch_offset = num_heads * sequence_length * head_size;
+  const int sequence_offset = num_heads * head_size;
+  const int head_offset = head_size;
+  bool has_bias = bias != nullptr;
+
+  TransferBSDToBNSHProgram program{"TransferBSDToBNSH", has_bias};
+  program.AddInputs({{input_tensor, ProgramTensorMetadataDependency::TypeAndRank}})
+      .AddOutputs({{output_tensor, ProgramTensorMetadataDependency::TypeAndRank}})
+      .SetDispatchGroupSize((data_size + WORKGROUP_SIZE - 1) / WORKGROUP_SIZE)
+      .AddUniformVariables({
+          {data_size},
+          {static_cast<uint32_t>(batch_offset)},
+          {static_cast<uint32_t>(sequence_offset)},
+          {static_cast<uint32_t>(head_offset)},
+          {static_cast<uint32_t>(bias_offset)}
+      });
+
+  if (has_bias) {
+    program.AddInput({bias, ProgramTensorMetadataDependency::TypeAndRank});
+  }
+
+  return context.RunProgram(program);
+};
+
+Status AttentionProbsProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  shader.AddInput("q", ShaderVariable::UseUniform | ShaderVariable::UseValueTypeAlias);
+  shader.AddInput("key", ShaderVariable::UseUniform | ShaderVariable::UseValueTypeAlias);
+  if (feed_past_key_) {
+    shader.AddInput("past_key", ShaderVariable::UseUniform);
+  }
+  if (has_attention_bias_) {
+    shader.AddInput("attention_bias", ShaderVariable::UseUniform);
+  }
+
+  shader.AddOutput("output", ShaderVariable::UseUniform | ShaderVariable::UseValueTypeAlias);
+  if (has_present_key_) {
+    shader.AddOutput("present_key", ShaderVariable::UseUniform);
+  }
+
+  shader.AppendImplementation("const TILE_SIZE = ", tile_size_, "u;\n")
```

Comment:
it seems that cache keys for the programs are not set correctly.

for example, here `tile_size_` is used as a part of the shader source code, but it is not set in the cache key. Use `program.CacheHint()` to set the cache key.

Reply:
TILE_SIZE can also be declared in overridable constants.

Reply:
Done